src/data_structures/suffix_array.rs

```diff
@@ -19,12 +20,79 @@ use vec_map::VecMap;
 
 use alphabets::{Alphabet, RankTransform};
 use data_structures::smallints::SmallInts;
+use data_structures::bwt::{BWT, Less, Occ};
 
+pub trait SuffixArray {
+    fn get(&self, index: usize) -> Option<usize>;
+    fn range(&self, range: Range<usize>) -> Option<Vec<usize>>;
```

Comment:
Strange. I thought it might work because I saw [this](http://stackoverflow.com/questions/33704592/performance-comparison-of-a-vec-and-a-boxed-slice). There, a box takes a slice as well. While ``[usize]`` is unsized, I believe ``&[usize]`` is not, which is why the example linked above works I guess. But that still does not solve the deref issue, or why box does not take the slice together with the reference as sized.

While your arguments for enum are also reasonable, I somehow come to the conclusion that this is also somehow a hint for us the sampled and the full suffix array are not similar enough to share a range method. Especially, looking at range in the sampled case shows me that it is not really performant, since a vector has to be created. Instead, it would be much more reasonable if SampledSuffixArray would return an iterator for a given range. Then, it is up to the user if a new array should be created or not. Imagine the interval of interest spans the whole or most of the array. If SampledSuffixArray was chosen to avoid memory issues, returning a vector of the same length as the original suffix array is not what people want.

Maybe this is what is common between both: returning an iterator for a given range. RawSuffixArray could have an additional function range_slice then, which returns the slice, but is not part of the common interface.